python/cudf/cudf/core/dataframe.py

```diff
@@ -2184,6 +2184,84 @@ def melt(self, **kwargs):
 
         return melt(self, **kwargs)
 
+    def _typecast_before_merge(self, lhs, rhs, left_on, right_on, how):
+        def casting_rules(dtype_l, dtype_r, how):
+            cast_warn = "can't safely cast column {} from {} with type \
+                         {} to {}, upcasting to {}"
+            ctgry_err = "can't implicitly cast column {} to categories \
+                         from right during left join"
+
+            rtn = None
+            if pd.api.types.is_dtype_equal(dtype_l, dtype_r):
+                rtn = dtype_l
+            elif how == "left":
+
+                check_col = rhs._cols[rcol][rhs._cols[rcol].notna()]
+                if not (
+                    Series(check_col) == Series(check_col).astype(dtype_l)
+                ).all():
+                    rtn = casting_rules(dtype_l, dtype_r, "inner")
+                    warnings.warn(
+                        cast_warn.format(rcol, "right", dtype_r, dtype_l, rtn)
+                    )
+                else:
+                    rtn = dtype_l
+            elif how == "right":
+                check_col = lhs._cols[lcol][lhs._cols[lcol].notna()]
+                if not (
+                    Series(check_col) == Series(check_col).astype(dtype_r)
+                ).all():
+                    rtn = casting_rules(dtype_l, dtype_r, "inner")
+                    warnings.warn(
+                        cast_warn.format(lcol, "left", dtype_l, dtype_r, rtn)
+                    )
+                else:
+                    rtn = dtype_r
+
+            elif is_categorical_dtype(dtype_l):
+                if how == "right":
+                    raise ValueError(ctgry_err.format(rcol))
```

Comment:
This will raise an error about a left join during a right join.

Reply:
@brandon-b-miller actually this still needs to be addressed before merge, super minor.